File: internal/protocol/codec/api/delete_groups.go
package api

import "github.com/woestebanaan/skafka/internal/protocol/codec"

// DeleteGroupsRequest (key 42, v0–v2). Used by AdminClient
// .deleteConsumerGroups() and kafka-consumer-groups.sh --delete to
// drop a consumer group's coordinator-side state plus its committed
// offsets. v2 introduces flexible (KIP-482 tagged-fields) framing.
type DeleteGroupsRequest struct {
	GroupNames []string
}

// DeleteGroupsResponse (key 42, v0–v2).
type DeleteGroupsResponse struct {
	ThrottleTimeMs int32
	Results        []DeleteGroupsResult
}

type DeleteGroupsResult struct {
	GroupID   string
	ErrorCode int16
}

func DecodeDeleteGroupsRequest(r *codec.Reader, version int16) (*DeleteGroupsRequest, error) {
	req := &DeleteGroupsRequest{}
	flexible := version >= 2

	readGroup := func() error {
		name, err := readString(r, flexible)
		if err != nil {
			return err
		}
		req.GroupNames = append(req.GroupNames, name)
		return nil
	}
	if flexible {
		if err := r.ReadCompactArray(readGroup); err != nil {
			return nil, err
		}
	} else {
		if err := r.ReadArray(readGroup); err != nil {
			return nil, err
		}
	}
	if flexible {
		if err := r.ReadTaggedFields(); err != nil {
			return nil, err
		}
	}
	return req, nil
}

func EncodeDeleteGroupsResponse(w *codec.Writer, resp *DeleteGroupsResponse, version int16) {
	flexible := version >= 2
	w.WriteInt32(resp.ThrottleTimeMs)

	writeResult := func() {
		for _, r := range resp.Results {
			writeString(w, r.GroupID, flexible)
			w.WriteInt16(r.ErrorCode)
			if flexible {
				w.WriteEmptyTaggedFields()
			}
		}
	}
	if flexible {
		w.WriteCompactArray(len(resp.Results), writeResult)
	} else {
		w.WriteArray(len(resp.Results), writeResult)
	}
	if flexible {
		w.WriteEmptyTaggedFields()
	}
}

File: crates/sk-codec/src/api/delete_groups.rs
//! DeleteGroups — API key 42.
//!
//! Versions 0..=2. Flexible (KIP-482) from v2. Used by
//! `AdminClient.deleteConsumerGroups()` and
//! `kafka-consumer-groups.sh --delete` to drop a consumer group's
//! coordinator-side state plus its committed offsets (gh #89).

use bytes::BytesMut;

use crate::api::common::{read_array_len, read_str, write_array_len, write_str};
use crate::api::registry::ApiSpec;
use crate::errors::CodecError;
use crate::headers::HeaderVersion;
use crate::primitives::{read_i16, read_i32, write_i16, write_i32};
use crate::tagged;
use crate::Bytes;

pub const VERSIONS: (i16, i16) = (0, 2);
pub const MIN_FLEXIBLE: i16 = 2;

const fn header_for(version: i16) -> HeaderVersion {
    if version >= MIN_FLEXIBLE {
        HeaderVersion::V2
    } else {
        HeaderVersion::V1
    }
}

fn request_hdr(version: i16) -> HeaderVersion {
    header_for(version)
}

fn response_hdr(version: i16) -> HeaderVersion {
    header_for(version)
}

pub const SPEC: ApiSpec = ApiSpec {
    key: 42,
    min_version: VERSIONS.0,
    max_version: VERSIONS.1,
    min_flexible: Some(MIN_FLEXIBLE),
    request_hdr,
    response_hdr,
};

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Request {
    pub group_names: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Response {
    pub throttle_time_ms: i32,
    pub results: Vec<DeleteGroupsResult>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DeleteGroupsResult {
    pub group_id: String,
    pub error_code: i16,
}

pub fn decode_request(buf: &mut Bytes, version: i16) -> Result<Request, CodecError> {
    let flexible = version >= MIN_FLEXIBLE;
    let mut req = Request::default();
    let n = read_array_len(buf, flexible)?;
    for _ in 0..n {
        req.group_names.push(read_str(buf, flexible)?);
    }
    if flexible {
        tagged::read(buf)?;
    }
    Ok(req)
}

pub fn encode_response(
    buf: &mut BytesMut,
    resp: &Response,
    version: i16,
) -> Result<(), CodecError> {
    let flexible = version >= MIN_FLEXIBLE;
    write_i32(buf, resp.throttle_time_ms);
    write_array_len(buf, resp.results.len(), flexible)?;
    for r in &resp.results {
        write_str(buf, &r.group_id, flexible)?;
        write_i16(buf, r.error_code);
        if flexible {
            tagged::write_empty(buf);
        }
    }
    if flexible {
        tagged::write_empty(buf);
    }
    Ok(())
}

pub fn decode_response(buf: &mut Bytes, version: i16) -> Result<Response, CodecError> {
    let flexible = version >= MIN_FLEXIBLE;
    let mut resp = Response {
        throttle_time_ms: read_i32(buf)?,
        ..Response::default()
    };
    let n = read_array_len(buf, flexible)?;
    for _ in 0..n {
        let group_id = read_str(buf, flexible)?;
        let error_code = read_i16(buf)?;
        if flexible {
            tagged::read(buf)?;
        }
        resp.results.push(DeleteGroupsResult {
            group_id,
            error_code,
        });
    }
    if flexible {
        tagged::read(buf)?;
    }
    Ok(resp)
}

pub fn encode_request(buf: &mut BytesMut, req: &Request, version: i16) -> Result<(), CodecError> {
    let flexible = version >= MIN_FLEXIBLE;
    write_array_len(buf, req.group_names.len(), flexible)?;
    for n in &req.group_names {
        write_str(buf, n, flexible)?;
    }
    if flexible {
        tagged::write_empty(buf);
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_request() -> Request {
        Request {
            group_names: vec!["g1".to_owned(), "g2".to_owned()],
        }
    }

    fn sample_response() -> Response {
        Response {
            throttle_time_ms: 0,
            results: vec![
                DeleteGroupsResult {
                    group_id: "g1".to_owned(),
                    error_code: 0,
                },
                DeleteGroupsResult {
                    group_id: "g2".to_owned(),
                    error_code: 69, // GROUP_ID_NOT_FOUND
                },
            ],
        }
    }

    fn roundtrip(version: i16) {
        let req = sample_request();
        let mut w = BytesMut::new();
        encode_request(&mut w, &req, version).unwrap();
        let mut r = w.freeze();
        let got = decode_request(&mut r, version).unwrap();
        assert_eq!(got, req);
        assert!(r.is_empty());

        let resp = sample_response();
        let mut w = BytesMut::new();
        encode_response(&mut w, &resp, version).unwrap();
        let mut r = w.freeze();
        let got = decode_response(&mut r, version).unwrap();
        assert_eq!(got, resp);
        assert!(r.is_empty());
    }

    #[test]
    fn v0_roundtrip() {
        roundtrip(0);
    }

    #[test]
    fn v1_roundtrip() {
        roundtrip(1);
    }

    #[test]
    fn v2_is_flexible() {
        roundtrip(2);
    }
}
